### PR TITLE
fix(providers): close Windows Antigravity and Grok gaps

### PR DESCRIPTION
**Pull request by Nanako0129**
## Summary

- Discover Antigravity's embedded Google OAuth client from the evidence-backed Windows 2.3.1 installation artifact while preserving the existing macOS candidates.
- Accept Google client IDs that are prefixed inside the Windows binary token span without weakening the existing ID grammar, pairing, ordering, or deduplication rules.
- Decode Grok's unified-billing included quota and on-demand cap while retaining legacy subscription precedence and independent window semantics.
- Accept both bare JSON numbers and `{ "val": number }` for unified Grok amounts, distinguish disabled quota pairs from malformed evidence, recognize a zero on-demand cap without requiring usage, and isolate synthesized monthly cards from unrecognized nested period evidence.
- Surface only sanitized provider source categories in Smoke output and update the provider-runtime validation matrix with the completed Windows evidence.

## Data flow and runtime behavior

Antigravity now resolves Windows candidates from `%LOCALAPPDATA%/Programs/antigravity/resources/bin/language_server.exe`, with the equivalent platform-home fallback when `LOCALAPPDATA` is unavailable. The scanner searches within the binary token span for a grammar-valid numeric Google client-ID start, then uses the existing client-ID/secret pairing policy. Embedded client values are neither logged nor persisted.

Grok keeps the legacy subscription order: matching `GrokBuild.productUsage.usagePercent` evidence wins, including zero, and a valid nonzero `creditUsagePercent` remains the fallback when no matching product percentage is supplied. Every matching GrokBuild row is scanned before accepting the first valid value; any explicitly malformed matching row fails closed regardless of duplicate order. Explicit-invalid legacy evidence is tracked independently and cannot be masked by disabled on-demand billing.

An exact raw zero `creditUsagePercent` is treated as lower-confidence only when a valid `used / monthlyLimit` pair exists. A strict exact-zero `used` plus exact-zero `monthlyLimit` pair is explicitly disabled and creates no included card; positive usage against a zero limit remains invalid. With an independently recognized zero on-demand cap, a fully disabled unified account succeeds with zero windows. Absent or malformed unified evidence otherwise preserves a valid legacy zero, while matching GrokBuild zero and valid nonzero legacy credit retain precedence.

Unified amounts share one strict parser that accepts a bare JSON number or `{ "val": number }`. Numeric strings, booleans, nulls, arrays, missing `val`, negative values, overflow/non-finite values, and syntactically nonzero values that underflow to signed zero fail closed. When legacy evidence is absent, `used / monthlyLimit` becomes the recurring included-quota percentage and defaults to the stable `billing.monthly.v1` card when the provider supplies no recognized period type.

Nested `currentPeriod.start/end` evidence is used only when `currentPeriod.type` maps to the existing weekly or monthly kinds. A synthesized monthly included card ignores unrecognized nested periods and uses only `billingPeriodStart/billingPeriodEnd`, preventing a daily or future period type from entering monthly pace/history. Present malformed fields on a recognized period still cannot be hidden by billing-level fallback.

A finite nonnegative on-demand pair with a positive cap produces `extra_usage.v1`, clamps over-cap use to 100%, uses the billing-period end only as display reset evidence, and marks pace as `nonRecurring` so the card never enters recurring history. Zero cap is recognized as disabled before reading on-demand usage, so missing or null `onDemandUsed` does not invalidate an exact-zero cap. Positive caps still require strict valid nonnegative usage; malformed included evidence cannot be hidden by a disabled cap, while independently valid positive Extra usage remains available. `prepaidBalance`, `topUpMethod`, and `isUnifiedBillingUser` are not treated as quota.

## Windows live validation

Exact head `6f99d3c6543b2889505521f111d19b0d399e74a3` passed the x64 `BuildTbNative` path, managed Smoke build, and credential-bound live poll:

- Codex: `oauth:2win`
- Claude: `oauth:2win`
- Antigravity: `cli:8win`
- Copilot: `oauth:2win`
- Grok: `oauth:0win` (successful recognized-disabled unified-billing result, not `response_shape`)

The installed Antigravity 2.3.1 `language_server.exe` passed a Windows-only temporary discovery assertion against the committed scanner. The final live poll retained identical exists/hash state for all seven monitored credential and account-scope paths. Antigravity remote OAuth success is intentionally not claimed because this Windows profile has no `~/.gemini/oauth_creds.json`; the remote flow remains covered hermetically.

## Verification

Latest exact head:

- `cargo test -p tb_core_ffi --release agent_grok::tests` — 42 passed
- `cargo test -p tb_core_ffi --release` — 287 passed
- `scripts/check.sh` — Rust workspace checks/tests, native build, 275 .NET tests, and Smoke passed
- Windows x64 `BuildTbNative` + managed Smoke + credential-bound live poll — passed
- Seven monitored credential/account-scope paths — unchanged before/after the live poll
- `git diff --check` — passed
- Fresh adversarial verifier — `CONFIRMED`

The local macOS host does not have the Windows ARM64 Rust target and Windows SDK/`clang-cl`; the latest-head `arm64-cross` GitHub job remains the required ARM64 build/package gate for this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/tb_core_ffi/src/agent_grok.rs [hunk 1]
```diff
@@ -233,45 +237,58 @@ fn map_billing(
         .config
         .ok_or_else(|| "Grok billing response missing config.".to_string())?;
 
-    let used_percent = used_percent_from_config(&config).ok_or_else(|| {
-        "Grok billing response has no creditUsagePercent or GrokBuild usage.".to_string()
-    })?;
+    let used_percent = used_percent_from_config(&config);
+    let (on_demand_recognized, extra_window) = extra_usage_window(&config, now);
+    if used_percent.is_none() && !on_demand_recognized {
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply crates/tb_core_ffi/src/agent_grok.rs]: Added defensive `config.used.val` / `config.monthlyLimit.val` parsing as the recurring included-quota fallback when no legacy GrokBuild or `creditUsagePercent` evidence is supplied. Valid included usage now maps to the provider period or `billing.monthly.v1` before a disabled on-demand cap can yield zero windows. Explicit-invalid legacy evidence still does not fall through; malformed included evidence is not masked by cap=0, while an independently valid positive `extra_usage.v1` remains available. The focused release suite now covers 31 Grok cases.

File: crates/tb_core_ffi/src/agent_grok.rs [hunk 2]
```diff
@@ -233,45 +247,75 @@ fn map_billing(
...
-                window =
-                    window.with_provider_duration_evidence(now, true, invalid_provider);
-            } else if let (Some(end), Some(duration_seconds)) = (period.end, period.duration_seconds)
-            {
-                window = window.with_provider_duration_evidence(
+    let mut windows = Vec::with_capacity(2);
+    if let Some(used_percent) = used_percent {
+        let period = period_details(&config);
+        let kind = period.kind.or(
+            included_monthly_fallback.then_some(("Monthly", "billing.monthly.v1")),
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply crates/tb_core_ffi/src/agent_grok.rs:297]: No code change for these names. The captured live response supplies `currentPeriod.type/start/end`, and the cited current Grok integrations also read nested `currentPeriod`; `UsagePeriodType`, `UsagePeriodStart`, and `UsagePeriodEnd` are their internal struct fields, not top-level JSON keys. They separately read top-level `billingPeriodStart` / `billingPeriodEnd`, which `period_details` already uses for reset and duration evidence. Without response or implementation evidence for top-level `usagePeriod*` JSON fields, accepting those aliases would mask an unverified schema and could assign history identity from speculative data. Included `monthlyLimit` therefore defaults to `billing.monthly.v1` only when the documented nested period has no recognized type.

File: crates/tb_core_ffi/src/agent_grok.rs [hunk 3]
```diff
@@ -297,14 +341,86 @@ fn used_percent_from_config(config: &BillingConfig) -> Option<f64> {
...
+        UsageWindow::from_used_percent("Extra usage".to_string(), used_percent, reset, now)
+            .with_identity("extra_usage.v1", Some("extra_usage.v1".to_string()))
+            .with_non_recurring();
+    (true, Some(window))
+}
+
+fn raw_amount_value(raw: &RawValue) -> Option<f64> {
+    let fields: std::collections::BTreeMap<String, Box<RawValue>> =
+        serde_json::from_str(raw.get()).ok()?;
+    let raw_value = fields.get("val")?.get();
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply crates/tb_core_ffi/src/agent_grok.rs]: `47d4023` now parses every unified amount through one strict raw-value helper that accepts either a bare JSON number or `{ "val": number }`. Numeric strings, wrong JSON types, non-finite/overflow values, and syntactically nonzero values that underflow to signed zero fail closed. Focused tests cover bare and wrapped included-quota and Extra-usage amounts.

File: crates/tb_core_ffi/src/agent_grok.rs [hunk 4]
```diff
@@ -233,45 +247,75 @@ fn map_billing(
...
 
-    let used_percent = used_percent_from_config(&config).ok_or_else(|| {
-        "Grok billing response has no creditUsagePercent or GrokBuild usage.".to_string()
-    })?;
+    let (used_percent, included_monthly_fallback, included_invalid) =
+        match legacy_percentage_evidence(&config) {
+            PercentageEvidence::Valid(percent) => (Some(percent), false, false),
+            PercentageEvidence::Invalid => (None, false, false),
+            PercentageEvidence::Absent => match included_percentage_evidence(&config) {
+                PercentageEvidence::Valid(percent) => (Some(percent), true, false),
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply crates/tb_core_ffi/src/agent_grok.rs]: `47d4023` treats only an exact raw zero `creditUsagePercent` as lower-confidence when a valid unified included pair exists. Matching `GrokBuild` zero and nonzero legacy credit retain precedence; absent or malformed unified evidence preserves the valid legacy zero. Tests cover each precedence branch and the floating-point underflow edge case.

File: crates/tb_core_ffi/src/agent_grok.rs [hunk 5]
```diff
@@ -233,45 +255,84 @@ fn map_billing(
...
+            | LegacyPercentageEvidence::Credit(percent) => (Some(percent), false, false),
+            LegacyPercentageEvidence::CreditZero(percent) => {
+                match included_percentage_evidence(&config) {
+                    PercentageEvidence::Valid(included) => (Some(included), true, false),
+                    PercentageEvidence::Absent | PercentageEvidence::Invalid => {
+                        (Some(percent), false, false)
+                    }
+                }
+            }
+            LegacyPercentageEvidence::Invalid => (None, false, false),
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply crates/tb_core_ffi/src/agent_grok.rs]: Final head `88bb72d` tracks explicit legacy invalidity independently from unified included and on-demand recognition, so malformed `GrokBuild.usagePercent` or `creditUsagePercent` returns the existing response-shape error even when the extra cap is disabled. It also scans every matching duplicate GrokBuild row; any malformed match fails closed regardless of order. Focused regressions cover both legacy sources and both duplicate orders.

File: crates/tb_core_ffi/src/agent_grok.rs [hunk 6]
```diff
@@ -233,45 +255,84 @@ fn map_billing(
...
-                window =
-                    window.with_provider_duration_evidence(now, true, invalid_provider);
-            } else if let (Some(end), Some(duration_seconds)) = (period.end, period.duration_seconds)
-            {
-                window = window.with_provider_duration_evidence(
+    let mut windows = Vec::with_capacity(2);
+    if let Some(used_percent) = used_percent {
+        let period = period_details(&config);
+        let kind = period.kind.or(
+            included_monthly_fallback.then_some(("Monthly", "billing.monthly.v1")),
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply crates/tb_core_ffi/src/agent_grok.rs:297]: Final head `88bb72d` uses nested `currentPeriod.start/end` only when the period type maps to the existing weekly or monthly kinds. Synthesized `billing.monthly.v1` cards ignore unrecognized nested periods and use only `billingPeriodStart/billingPeriodEnd`; recognized periods retain the existing malformed-primary fail-closed rule. Tests cover DAILY nested dates both with and without billing-level dates.

File: crates/tb_core_ffi/src/agent_grok.rs [hunk 7]
```diff
@@ -233,45 +255,84 @@ fn map_billing(
...
-    })?;
+    let (used_percent, included_monthly_fallback, included_invalid) =
+        match legacy_percentage_evidence(&config) {
+            LegacyPercentageEvidence::GrokBuild(percent)
+            | LegacyPercentageEvidence::Credit(percent) => (Some(percent), false, false),
+            LegacyPercentageEvidence::CreditZero(percent) => {
+                match included_percentage_evidence(&config) {
+                    PercentageEvidence::Valid(included) => (Some(included), true, false),
+                    PercentageEvidence::Absent | PercentageEvidence::Invalid => {
+                        (Some(percent), false, false)
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply crates/tb_core_ffi/src/agent_grok.rs]: Final head `88bb72d` classifies a strict exact-zero `used`/`monthlyLimit` pair as explicitly disabled rather than invalid. Exact-zero legacy credit then creates no included card, and an independently disabled on-demand pair yields a successful zero-window snapshot. Positive usage with a zero limit remains invalid, while GrokBuild zero and valid nonzero legacy credit retain precedence.

File: crates/tb_core_ffi/src/agent_grok.rs [hunk 8]
```diff
@@ -297,14 +363,107 @@ fn used_percent_from_config(config: &BillingConfig) -> Option<f64> {
...
+    PercentageEvidence::Valid(((used / monthly_limit) * 100.0).min(100.0))
+}
+
+fn extra_usage_window(config: &BillingConfig, now: DateTime<Utc>) -> (bool, Option<UsageWindow>) {
+    let Some(used) = config
+        .on_demand_used
         .as_deref()
-        .and_then(valid_percentage)
+        .and_then(raw_amount_value)
+    else {
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply crates/tb_core_ffi/src/agent_grok.rs]: Final head `6f99d3c` parses `onDemandCap` before `onDemandUsed`. A strict finite exact-zero cap is recognized as disabled even when usage is missing or null; positive caps still require strict valid nonnegative usage, while malformed, negative, non-finite, and underflowed caps remain fail-closed. Regressions cover missing usage, null usage, and the fully disabled 0/0/default-zero shape.
